### ISSUE TITLE
Reporting incorrect (pretty sure) CAPABILITY_NETWORK and CAPABILITY_ARBITRARY_EXECUTION

### ISSUE THREAD

**Issue by ldemailly** [google/capslock#144]
```
git clone https://github.com/fortio/log
cd log
# btw... consider moving the main at the top for shorter install cmd/ is overused
go install github.com/google/capslock/cmd/capslock@latest 
# ( got capslock v0.2.4 )
capslock --output=verbose
[...]
CAPABILITY_NETWORK: 7 references (7 direct, 0 transitive)
Example callpath:
  fortio.org/log.LogAndCall$1$1
  http_logging.go:202:15:fortio.org/log.LogRequest
  http_logging.go:84:56:(net/http.Header).Get
[...]
CAPABILITY_ARBITRARY_EXECUTION: 30 references (0 direct, 30 transitive)
Example callpath:
  fortio.org/log.init#1
  logger.go:190:15:fortio.org/log.configFromEnv
  logger.go:201:8:fortio.org/log.Errf
  logger.go:520:11:fortio.org/log.logPrintf
  logger.go:414:19:fortio.org/log.logUnconditionalf
  logger.go:447:31:fortio.org/log.colorGID
  console_logging.go:144:55:fortio.org/log/goroutine.ID
  gid.go:20:22:fortio.org/log/goroutine.add
```
https://github.com/fortio/log

doing http.Header.Get shouldn't be "CAPABILITY_NETWORK" should it ?

and the hacky code in https://github.com/fortio/log/blob/main/goroutine/gid.go#L20
is using unsafe but... how does that turn into CAPABILITY_ARBITRARY_EXECUTION ?

**Comment by jcd2** [comment]
Thanks for the report!

> doing http.Header.Get shouldn't be "CAPABILITY_NETWORK" should it ?

You're right, I'll make a change soon to add this (and some similar functions) to Capslock's list of functions under net/ that don't really have the network capability.

> how does that turn into CAPABILITY_ARBITRARY_EXECUTION ?

It's the call on that line to a go:linkname function that gets it marked as arbitrary execution.  Capslock isn't clever enough to follow go:linkname directives, but we'll look into adding that in the future.

**Comment by ldemailly** [comment]
ping?

**Comment by jcd2** [comment]
We've narrowed the set of functions in `net` that have capabilities, but I'm afraid we haven't done any work on resolving the effects of linkname yet, which would probably be implemented upstream of Capslock in golang.org/x/tools.  I'll retitle this issue to cover that part.

**Comment by ldemailly** [comment]
Thanks for the reply,

> We've narrowed the set of functions in net that have capabilities

I still get
```
CAPABILITY_NETWORK: 6 references (6 direct, 0 transitive)
Example callpath:
  fortio.org/log.LogAndCall$1
  http_logging.go:189:4:fortio.org/log.LogAndCall$1$1
  http_logging.go:197:17:net/http.Error
```

(maybe because of init/globals in net/http? but should callers of http.Header.Get get that label?)

**Comment by jcd2** [comment]
> 
> I still get
> 
> ```
> CAPABILITY_NETWORK: 6 references (6 direct, 0 transitive)
> Example callpath:
>   fortio.org/log.LogAndCall$1
>   http_logging.go:189:4:fortio.org/log.LogAndCall$1$1
>   http_logging.go:197:17:net/http.Error
> ```
> 

Well, `net/http.Error` here writes an error response that gets sent over the network:

[https://github.com/fortio/log/blob/main/http_logging.go#L197](https://github.com/fortio/log/blob/main/http_logging.go#L197)

so IMO it should have the network capability.

**Comment by ldemailly** [comment]
Indeed I hadn't notice which method is flagged has changed, yet http.Error imo isn't network capabilites, it doesn't do anything if you don't have a prexisting http.Response - network capability would be imo either listening to a port or opening a new connection; neither of which does that code do

**Comment by jcd2** [comment]
That's true, it's only affecting a network connection that already exists.

We've been thinking about adding finer-grained capabilities, and your case here would be a good one for us to improve on: distinguishing between opening or listening for new connections, and reading or writing data.

**Comment by ldemailly** [comment]
thanks, would io.Write be flagged as “network capabilities” though ? it shouldn’t imo

**Comment by jcd2** [comment]
> thanks, would io.Write be flagged as “network capabilities” though ? it shouldn’t imo

If the Write call sends bytes over an existing network connection, I think it should.  But ideally it would be a different, lesser capability than opening a new connection.
